pkg/traceql/ast.go

```diff
@@ -285,15 +287,22 @@ func (f SpansetFilter) evaluate(input []*Spanset) ([]*Spanset, error) {
 				continue
 			}
 
-			matchingSpans = append(matchingSpans, s)
+			f.matchingSpansBuffer = append(f.matchingSpansBuffer, s)
+		}
+
+		if len(f.matchingSpansBuffer) == 0 {
+			continue
 		}
 
-		if len(matchingSpans) == 0 {
+		if len(f.matchingSpansBuffer) == len(ss.Spans) {
```

Comment:
so, i think this is safe, but i need to mention the bug where manipulating the input slice of spans caused the double add the sync pool and panic'ed everything.

we do have tests this cover this and working through this in my head makes me believe there are no issues.